### PR TITLE
Show error icon on Nuimo when something goes wrong with calling Home Assistant

### PR DESCRIPTION
**Pull request by grunskis**
In the process refactored the the way how websocket communications are done to make this happen. The refactoring was also necessary to be able to better time requests for smoother brightness / volume update that I'll work on next...

### CODE REVIEW

File: senic_hub/nuimo_app/hass.py [hunk 1]
```diff
@@ -6,13 +6,13 @@
 from pprint import pformat
 from threading import Thread
 
-from websocket import WebSocketConnectionClosedException, create_connection
+import websocket
```
Comment (tomster) [inline senic_hub/nuimo_app/hass.py]: was this meant to be part of this commit?

File: senic_hub/nuimo_app/hass.py [hunk 2]
```diff
@@ -29,116 +84,106 @@ def __init__(self, url, ws_connection=None, on_connect=None, on_disconnect=None)
...
         self.stopping = False
-        self.response_callbacks = {}  # req_id: callback
 
+        self.result_callbacks = {}  # req_id: callback
         self.state_listeners = defaultdict(list)  # entity_id: [callback1, ...]
 
-    def connect(self, reconnect=False, reconnect_timeout=5):
-        self.connection = None
+    def connect(self):
+        self.connection = HomeAssistantConnection(
```
Comment (tomster) [inline senic_hub/nuimo_app/hass.py]: does this mean, that the connection instance is now created adhoc, IOW on demand? why was this change necessary or what is the benefit? (just trying to understand)

File: senic_hub/nuimo_app/hass.py [hunk 3]
```diff
@@ -29,116 +84,106 @@ def __init__(self, url, ws_connection=None, on_connect=None, on_disconnect=None)
...
-            if not result:
-                return
-
-            if not (result["id"] == request["id"] and
-                    result["type"] == "result" and
-                    result["success"] is True):
-
-                logger.error(result)
-                return None
+            if self.connection and self.connection.connected:
```
Comment (tomster) [inline senic_hub/nuimo_app/hass.py]: nice!

File: senic_hub/nuimo_app/tests/test_hass.py [hunk 4]
```diff
@@ -1,17 +1,15 @@
...
-from unittest.mock import MagicMock
 
 from senic_hub.nuimo_app.hass import HomeAssistant
 
 
 class TestHomeAssistant(TestCase):
     def test_find_state_entity_state(self):
-        connection = MagicMock()
-        listener = HomeAssistant("ws://localhost:8123", connection)
+        ha = HomeAssistant("ws://localhost:8123")
```
Comment (tomster) [inline senic_hub/nuimo_app/tests/test_hass.py]: does this mean, that test runs require a running HA instance? this commit seems to only remove the mocking without replacing it with something else

File: senic_hub/nuimo_app/hass.py (hunk 3 above)
Comment (tomster) [inline senic_hub/nuimo_app/hass.py]: i know it's not a big deal, i'm just a sucker for these little things...

File: senic_hub/nuimo_app/tests/test_hass.py (hunk 4 above)
Comment (grunskis) [inline senic_hub/nuimo_app/tests/test_hass.py]: The connection is not created in constructor anymore so there's no need for the mock. The URL is required thought, but I could probably replace it with some dummy one...

File: senic_hub/nuimo_app/hass.py (hunk 1 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: Yep, it's legit. The websocket library supports 2 ways of using it - the simple one (it's what I used before) and async one (based on callbacks) which I'm using now

File: senic_hub/nuimo_app/hass.py (hunk 2 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: Not really. In fact I'm trying to always have connection available if possible (if Home Assistant is up and running).

Does this answer the question? Maybe I misunderstood what you were asking...

File: senic_hub/nuimo_app/hass.py (hunk 3 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: Didn't get this one. Looks like something is missing in your comment...

File: senic_hub/nuimo_app/tests/test_hass.py (hunk 4 above)
Comment (grunskis) [inline senic_hub/nuimo_app/tests/test_hass.py]: Addressed here 6308663

File: senic_hub/nuimo_app/hass.py [hunk 5]
```diff
@@ -1,24 +1,79 @@
...
+    def __init__(self, url, on_open, on_message, on_close):
+        super().__init__()
+
+        self.ws_url = "ws://{}/api/websocket".format(url)
+
+        self.on_open_callback = on_open
+        self.on_message_callback = on_message
+        self.on_close_callback = on_close
+
+        self.ws = None
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: I would name it `websocket`

File: senic_hub/nuimo_app/hass.py [hunk 6]
```diff
@@ -1,24 +1,79 @@
...
 
 logger = logging.getLogger(__name__)
 
 
-class HAListener(Thread):
+class HomeAssistantConnection(Thread):
+    def __init__(self, url, on_open, on_message, on_close):
+        super().__init__()
+
+        self.ws_url = "ws://{}/api/websocket".format(url)
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: I would name it `websocket_url`

File: senic_hub/nuimo_app/hass.py [hunk 7]
```diff
@@ -1,24 +1,79 @@
...
+        self.on_open_callback()
+
+    def on_message(self, ws, message):
+        self.on_message_callback(json.loads(message))
+
+    def on_error(self, ws, error):
+        logger.error("Connection failed with: %s", error)
+        self.ws = None
+
+    def on_close(self, ws):
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: I would make those methods "private" (prefix with `_`) so the user of this class does not accidentally assign his callbacks to `home_assistant.on_close` instead of `home_assistant.on_close_callback`

File: senic_hub/nuimo_app/hass.py [hunk 8]
```diff
@@ -29,116 +84,106 @@ def __init__(self, url, ws_connection=None, on_connect=None, on_disconnect=None)
...
-            logger.warn("Home Assistant connection lost! Trying to reconnect...")
+        self.next_reconnect_time = time() + self.reconnection_interval
 
-            on_disconnect_listener = self.connection_state_listeners["on_disconnect"]
-            if on_disconnect_listener:
-                on_disconnect_listener()
+    def on_connect(self):
+        on_connect_listener = self.connection_state_listeners["on_connect"]
+        if on_connect_listener:
+            on_connect_listener()
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: Can be made functional by writing:
```python
(self.connection_state_listeners["on_connect"] or (lambda: None))()
```

File: senic_hub/nuimo_app/hass.py [hunk 9]
```diff
@@ -29,116 +84,106 @@ def __init__(self, url, ws_connection=None, on_connect=None, on_disconnect=None)
...
-        except (ConnectionResetError, WebSocketConnectionClosedException):
-            self.connect(reconnect=True)
+    def on_disconnect(self):
+        if self.connection:
+            self.connection.stop()
+            self.connection = None
+
+        on_disconnect_listener = self.connection_state_listeners["on_disconnect"]
+        if on_disconnect_listener:
+            on_disconnect_listener()
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: Same functional craziness could be applied here 🐵

File: senic_hub/nuimo_app/hass.py [hunk 10]
```diff
@@ -29,116 +84,106 @@ def __init__(self, url, ws_connection=None, on_connect=None, on_disconnect=None)
...
+            self.url,
+            on_open=self.on_connect,
+            on_message=self.on_message,
+            on_close=self.on_disconnect,
+        )
+        self.connection.start()
 
-        if reconnect:
-            logger.warn("Home Assistant connection lost! Trying to reconnect...")
+        self.next_reconnect_time = time() + self.reconnection_interval
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: I would only remember when the connection was started and then later check if enough time has passed by to try a reconnection. I find that approach better because the implementation for what a "reconnection time" means is not encoded here but actually in the place where it's needed, namely line 181.

File: senic_hub/nuimo_app/hass.py [hunk 11]
```diff
@@ -29,116 +84,106 @@ def __init__(self, url, ws_connection=None, on_connect=None, on_disconnect=None)
...
-            elif result["type"] == "result":
-                if self.process_response_callbacks(result):
-                    continue
-                else:
-                    logger.debug("Got result w/o callback: %s", result)
-            else:
-                logger.debug("Unknown HA message: %s", result)
+            if not (self.connection and self.connection.connected):
+                if time() >= self.next_reconnect_time:
+                    self.connect()
```
Comment (larsblumberg) [inline senic_hub/nuimo_app/hass.py]: 1. Given the above comment (remembering start time of connection), I would type here: if time() >= self.connection_time + self.reconnection_interval:

2. Why not combining both if statements into a single if statement with and-joined conditions?

File: senic_hub/nuimo_app/hass.py (hunk 3 above)
Comment (tomster) [inline senic_hub/nuimo_app/hass.py]: it was just a comment on my comment :)

File: senic_hub/nuimo_app/hass.py (hunk 7 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: The callbacks are passed in as parameters to the init method and not assigned directly.

File: senic_hub/nuimo_app/hass.py (hunk 8 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: I would rather do this:

```
def noop():
    pass

...
    def __init__(self, url, on_connect=noop, on_disconnect=noop):
        super().__init__()

        self.connection_state_listeners = {
            "on_connect": on_connect,
            "on_disconnect": on_disconnect,
        }

...

on_disconnect_listener = self.connection_state_listeners["on_disconnect"]
on_disconnect_listener()
```

But I'm not sure if this is not making the code too smart.

File: senic_hub/nuimo_app/hass.py (hunk 6 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: Renamed ca99fa1

File: senic_hub/nuimo_app/hass.py (hunk 11 above)
Comment (grunskis) [inline senic_hub/nuimo_app/hass.py]: Discussed this IRL with @larsblumberg and decided to leave it as is for now and observe how it behaves...
